topdown/eval.go

```diff
@@ -3197,32 +3245,60 @@ func (e evalTerm) next(iter unifyIterator, plugged *ast.Term) error {
 }
 
 func (e evalTerm) enumerate(iter unifyIterator) error {
-
+	var outerEe *outerEarlyExitError
 	switch v := e.term.Value.(type) {
 	case *ast.Array:
 		for i := 0; i < v.Len(); i++ {
 			k := ast.IntNumberTerm(i)
 			err := e.e.biunify(k, e.ref[e.pos], e.bindings, e.bindings, func() error {
 				return e.next(iter, k)
 			})
+
 			if err != nil {
-				return err
+				if oee, ok := err.(*outerEarlyExitError); ok {
+					if outerEe == nil {
+						outerEe = oee
+					}
+				} else {
+					return err
+				}
 			}
 		}
 	case ast.Object:
 		return v.Iter(func(k, _ *ast.Term) error {
-			return e.e.biunify(k, e.ref[e.pos], e.termbindings, e.bindings, func() error {
+			err := e.e.biunify(k, e.ref[e.pos], e.termbindings, e.bindings, func() error {
 				return e.next(iter, e.termbindings.Plug(k))
 			})
+			if oee, ok := err.(*outerEarlyExitError); ok {
+				if outerEe == nil {
+					outerEe = oee
+				}
+				return nil
+			}
+			return err
 		})
 	case ast.Set:
-		return v.Iter(func(elem *ast.Term) error {
-			return e.e.biunify(elem, e.ref[e.pos], e.termbindings, e.bindings, func() error {
+		if err := v.Iter(func(elem *ast.Term) error {
+			err := e.e.biunify(elem, e.ref[e.pos], e.termbindings, e.bindings, func() error {
 				return e.next(iter, e.termbindings.Plug(elem))
 			})
-		})
+			if oee, ok := err.(*outerEarlyExitError); ok {
+				if outerEe == nil {
+					outerEe = oee
+				}
+				return nil
+			}
+
+			return err
+		}); err != nil {
+			return err
+		}
 	}
 
+	// FIXME: What if we get multiple outerEarlyExitErrors?
+	if outerEe != nil {
+		return outerEe
+	}
 	return nil
```

Comment:
[nit] inconsequential, but this would be the same as
```suggestion
	return outerEe
```

Reply:
Unless you're doing type matching on the returned value, as the returned `nil` would still have the type  `earlyExitError`.